sgkit/stats/aggregation.py

```diff
@@ -1,10 +1,103 @@
 import dask.array as da
+import numba
 import numpy as np
 import xarray as xr
 from xarray import DataArray, Dataset
 
+from ..typing import ArrayLike
 
-def count_alleles(ds: Dataset) -> DataArray:
+
+@numba.guvectorize(  # type: ignore
+    [
+        "void(numba.int8[:], numba.uint8[:], numba.uint8[:])",
+        "void(numba.int16[:], numba.uint8[:], numba.uint8[:])",
+        "void(numba.int32[:], numba.uint8[:], numba.uint8[:])",
+        "void(numba.int64[:], numba.uint8[:], numba.uint8[:])",
+    ],
+    "(k),(n)->(n)",
+    nopython=True,
+)
+def count_alleles(g: ArrayLike, _: ArrayLike, out: ArrayLike) -> None:
+    """Generaliszed U-function for computing per sample allele counts.
+
+    Parameters
+    ----------
+    g : (K,) array-like, int
+        A genotype call with K alleles where K is the genotypes ploidy.
+    _: (N,) array-like, uint8
+        Dummy variable of length N where N is the number of possible
+        unique alleles.
+
+    Returns
+    -------
+    ac : (N,) array-like, uint8
+        Allele counts with values corresponding to the number of
+        non-missing occurrences of each allele whithin g.
+
+    """
+    out[:] = 0
+    n_allele = len(g)
+    for i in range(n_allele):
+        a = g[i]
+        if a >= 0:
+            out[a] += 1
+
+
+def count_call_alleles(ds: Dataset) -> DataArray:
+    """Compute per sample allele counts from genotype calls.
+
+    Parameters
+    ----------
+    ds : Dataset
+        Genotype call dataset such as from
+        `sgkit.create_genotype_call_dataset`.
+
+    Returns
+    -------
+    call_allele_count : DataArray
+        Allele counts with shape (variants, samples, alleles) and values
+        corresponding to the number of non-missing occurrences
+        of each allele.
+
+    Examples
+    --------
+
+    >>> import sgkit as sg
+    >>> from sgkit.testing import simulate_genotype_call_dataset
+    >>> ds = simulate_genotype_call_dataset(n_variant=4, n_sample=2, seed=1)
+    >>> ds['call_genotype'].to_series().unstack().astype(str).apply('/'.join, axis=1).unstack() # doctest: +NORMALIZE_WHITESPACE
+    samples 0   1
+    variants
+    0       1/0	1/0
+    1       1/0	1/1
+    2       0/1	1/0
+    3       0/0	0/0
+
+    >>> sg.count_call_alleles(ds).values # doctest: +NORMALIZE_WHITESPACE
+    array([[[1, 1],
+            [1, 1]],
+    <BLANKLINE>
+           [[1, 1],
+            [0, 2]],
+    <BLANKLINE>
+           [[1, 1],
+            [1, 1]],
+    <BLANKLINE>
+           [[2, 0],
+            [2, 0]]], dtype=uint8)
+    """
+    n_alleles = ds.dims["alleles"]
+    G = da.asarray(ds["call_genotype"])
+    shape = (G.chunks[0], G.chunks[1], n_alleles)
+    N = da.empty(n_alleles, dtype=np.uint8)
```

Comment:
Good idea using this instead!  For the subsequent map_blocks call it might be a good idea to ensure this has only a single chunk so that the blocks broadcast.  I can't imagine why anyone would ever configure the default chunk size to be so small that `n_alleles` items would result in multiple chunks, but guarding against it is probably a good idea.

Reply:
>  I can't imagine why anyone would ever configure the default chunk size to be so small that `n_alleles` items would result in multiple chunks

Neither but this edge case should be handled correctly by this PR, see below.

Reply:
Ah I see.  Thoughts on a `da.asarray(ds["call_genotype"].chunk(chunks=dict(ploidy=-1)))` then?  I think ploidy chunks of size one could actually be pretty common when concatenating haplotype arrays.

I could also see the argument in leaving it up to the user to interpret those error messages and rechunk themselves so that they have to think through the performance implications, but I had already started being defensive about that in the GWAS regression methods.  My rationale was that it's better to optimize for a less frustrating user experience than for making performance concerns prominent (and we could always add warnings for that later).

Reply:
Ohh nm, I see why it doesn't matter now (and the test for it).  Ignore that.